Azure/data-api-builder
# Adding local CORS support

Closes #305
## Why is this change being made?
See #305. Briefly, CORS support is needed if we want to be able to interact with hawaii from an external/3rd party client. For example, I am using Apollo GraphQL in a Next.js application as my client to interact with project hawaii through gql queries. As @sajeetharan mentions in the issue, in production this won't be relevant as CORS can be configured through AppService or SWAs. However, while hawaii is still in development, this is necessary for simulating client interaction. 

The status quo is as follows: on sending an HTTP request to hawaii's graphql endpoint (`http://localhost:5001/graphql`) from a given client origin (`http://localhost:3000`), we are greeted with the infamous `blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource` in the developer console.

### Why isn't this an issue with Postman? 
One may wonder why Postman queries work without any CORS configuration, seeing as it is also a REST client. Note that CORS is a server-side configuration that **is up to clients to enforce**.  Browsers usually enforce CORS rules because of the risk of Cross-Site Request Forgery (CSRF) attacks, where an attacker may use browser credentials, such as cookies in session storage, to make unauthorized cross origin requests. However, developer tools like Postman and cURL don't include such risk, thus they choose to effectively ignore CORS headers returned by the server.

### What about Banana Cake Pop (BCP)?
BCP is a web interface that operates through the browser - why wasn't CORS config needed here? The simple answer is that theoretically, BCP makes requests from the same origin that it is issuing responses from (that being https://localhost:5001/graphql for example), thus it qualifies for the same-origin policy that bypasses CORS. Please see [Same-Origin Policy (SOP)](https://developer.mozilla.org/en-US/docs/Web/Security/Same-origin_policy) for more, and to see what distinguishes origins i.e. (scheme, domain, port) tuples.

## What changed?
The `origins` and `allow-credentials` fields from the runtime config are no longer ignored. They are read in the `Configure` method of `Startup.cs` and passed as options when configuring CORS middleware. No other services should be affected. The default values in all configuration files and example overrides have been updated to 
``` 
"origins": [ "http://localhost:3000" ],
"allow-credentials": false 
```
Unit tests related to CORS configuration have been added in a new testing file: 
`hawaii-gql\DataGateway.Service.Tests\Configuration\CorsUnitTests.cs`
## How was this validated?
 - Manual Testing
     - Both allowing all origins by using the "*" wildcard or specifying an individual origin "http://localhost:3000" result in `Access-Control-Allow-Origin` headers correctly returned by the server, and the Apollo client can freely interact with hawaii
     - Configuration of allowing all origins AND allowing credentials will throw an `InvalidOperationException`
         - this is an [insecure configuration](https://docs.microsoft.com/en-us/aspnet/core/security/cors?view=aspnetcore-6.0#:~:text=to%20the%20app.-,Note,-Specifying%20AllowAnyOrigin%20and) as the Microsoft docs point out, as it puts one at risk for CSRF attacks
     - Supplying the "*" wildcard **alone** in the origins list correctly allows all origins
         - Including any other element in the origins list will nullify the wildcard, as discovered through testing. Only reference found in Microsoft documentation is [here](https://docs.microsoft.com/en-us/cli/azure/webapp/cors?view=azure-cli-latest). Behavior may be somewhat unpredictable so could use formal documentation.
 - New CORS unit tests ensure:
     - Cors options are correctly deserialized and read in runtime config
     - Supplying "*" individually or including "http://localhost:3000" in the origins list correctly returns the `Access-Control-Allow-Origin` header on responses to HTTP requests with an Origin header set accordingly
     - Toggling the AllowCredentials option correctly toggles the `Access-Control-Allow-Credentials` header on responses to HTTP requests coming from valid origins
     - http vs https scheme is correctly distinguished (Allow-Origin header not present on response if the two are mismatched on origins list and client header)
     

## Middleware Order
One last small point to document: the order of middleware being added **is** relevant. See [middleware order](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/middleware/?view=aspnetcore-6.0#middleware-order). Basically, `useCors` call must come after `useRouting` but before `useAuthentication`.
